File: src/main/java/alfheim/common/entity/spell/EntitySpellFenrirStorm.java
package alfheim.common.entity.spell;

import java.util.List;
import java.util.UUID;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.OrientedBB;
import alexsocol.asjlib.math.Vector3;
import alfheim.AlfheimCore;
import alfheim.api.spell.ITimeStopSpecific;
import alfheim.api.spell.SpellBase;
import alfheim.common.core.util.DamageSourceSpell;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.world.World;

public class EntitySpellFenrirStorm extends Entity implements ITimeStopSpecific {
	
	public OrientedBB area;
	public EntityLivingBase caster;
	
	public EntitySpellFenrirStorm(World world) {
		super(world);
		setSize(0.1F, 0.1F);
		area = new OrientedBB(AxisAlignedBB.getBoundingBox(-0.5, -0.5, -8, 0.5, 0.5, 8));
		renderDistanceWeight = 4;
	}
	
	public EntitySpellFenrirStorm(World world, EntityLivingBase caster) {
		this(world);
		this.caster = caster;
		Vector3 l = new Vector3(caster.getLookVec()).mul(0.1);
		setPositionAndRotation(caster.posX + l.x, caster.posY + caster.getEyeHeight() + l.y, caster.posZ + l.z, caster.rotationYaw, caster.rotationPitch);
		
		area.translate(caster.posX, caster.posY + caster.getEyeHeight(), caster.posZ);
		area.rotateOX(caster.rotationPitch);
		area.rotateOY(-caster.rotationYaw);
		
		Vector3 v = new Vector3(caster.getLookVec()).mul(8.5);
		area.translate(v.x, v.y, v.z);
	}
	
	@Override
	public void onEntityUpdate() {
		if (!AlfheimCore.enableMMO || (!worldObj.isRemote && (caster == null || area == null || ticksExisted > 12))) {
			setDead();
			return;
		}
		if (this.isDead || !ASJUtilities.isServer()) return;
		
		if (ticksExisted == 4) {
			List<EntityLivingBase> l = worldObj.getEntitiesWithinAABB(EntityLivingBase.class, area.toAABB());
			for (EntityLivingBase e : l) if (e != caster && area.intersectsWith(e.boundingBox)) e.attackEntityFrom(DamageSourceSpell.lightning(this, caster), SpellBase.over(caster, 10));
		}
	}
	
	public boolean isImmune() {
		return false;
	}
	
	public boolean affectedBy(UUID uuid) {
		return !caster.getUniqueID().equals(uuid);
	}
	
	public void entityInit() {}
	
	public void readEntityFromNBT(NBTTagCompound nbt) {
		if (nbt.hasKey("castername")) caster = worldObj.getPlayerEntityByName(nbt.getString("castername")); else setDead();
		if (caster == null) setDead();
	}
	
	public void writeEntityToNBT(NBTTagCompound nbt) {
		if (caster instanceof EntityPlayer) nbt.setString("castername", caster.getCommandSenderName());
	}
}

File: src/main/java/alfheim/common/entity/spell/EntitySpellFenrirStorm.kt
package alfheim.common.entity.spell

import alexsocol.asjlib.*
import alexsocol.asjlib.math.*
import alfheim.api.spell.*
import alfheim.common.core.handler.AlfheimConfigHandler
import alfheim.common.core.util.DamageSourceSpell
import alfheim.common.spell.wind.SpellFenrirStorm
import net.minecraft.entity.*
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.AxisAlignedBB
import net.minecraft.world.World
import java.util.*

class EntitySpellFenrirStorm(world: World): Entity(world), ITimeStopSpecific {
	
	val area: OrientedBB
	var caster: EntityLivingBase? = null
	var mjolnir: Boolean
		get() = dataWatcher.getWatchableObjectInt(2) != 0
		set(value) = dataWatcher.updateObject(2, if (value) 1 else 0)
	
	override val isImmune: Boolean
		get() = false
	
	init {
		setSize(0.25f, 0.25f)
		area = OrientedBB(AxisAlignedBB.getBoundingBox(-0.5, -0.5, -SpellFenrirStorm.radius, 0.5, 0.5, SpellFenrirStorm.radius))
		renderDistanceWeight = SpellFenrirStorm.radius / 2
	}
	
	constructor(world: World, caster: EntityLivingBase, mjolnir: Boolean = false): this(world) {
		this.caster = caster
		val l = Vector3(caster.lookVec).mul(0.1)
		setPositionAndRotation(caster.posX + l.x, caster.posY + caster.eyeHeight.D + l.y, caster.posZ + l.z, caster.rotationYaw, caster.rotationPitch)
		
		area.translate(caster.posX, caster.posY + caster.eyeHeight, caster.posZ)
		area.rotateOX(-caster.rotationPitch.D)
		area.rotateOY((caster.rotationYaw).D)
		
		val v = Vector3(caster.lookVec).mul(SpellFenrirStorm.radius + 0.5)
		area.translate(v.x, v.y, v.z)
		
		this.mjolnir = mjolnir
	}
	
	override fun onEntityUpdate() {
		if ((!AlfheimConfigHandler.enableMMO && !mjolnir) || (!worldObj.isRemote && caster == null) || ticksExisted > 12) {
			setDead()
			return
		}
		if (isDead || ASJUtilities.isClient) return
		
		val caster = caster ?: return
		
		if (mjolnir) {
			rotationYaw = caster.rotationYaw
			rotationPitch = caster.rotationPitch
			
			area.fromAABB(AxisAlignedBB.getBoundingBox(-0.5, -0.5, -SpellFenrirStorm.radius, 0.5, 0.5, SpellFenrirStorm.radius))
			area.translate(caster.posX, caster.posY + caster.eyeHeight, caster.posZ)
			area.rotateOX(-caster.rotationPitch.D)
			area.rotateOY((caster.rotationYaw).D)
			
			val v = Vector3(caster.lookVec).mul(SpellFenrirStorm.radius + 0.5)
			area.translate(v.x, v.y, v.z)
		}
		
		boundingBox.setBB(area.toAABB())
		
		if (ticksExisted == 4 || mjolnir) {
			val l = worldObj.getEntitiesWithinAABB(EntityLivingBase::class.java, area.toAABB()) as List<EntityLivingBase>
			for (e in l)
				if (e !== caster && area.intersectsWith(e.boundingBox))
					e.attackEntityFrom(DamageSourceSpell.lightning(this, caster), SpellBase.over(caster, SpellFenrirStorm.damage.D))
		}
	}
	
	override fun affectedBy(uuid: UUID) = caster!!.uniqueID != uuid
	
	public override fun entityInit() {
		dataWatcher.addObject(2, 0)
	}
	
	public override fun readEntityFromNBT(nbt: NBTTagCompound) {
		if (nbt.hasKey("castername")) caster = worldObj.getPlayerEntityByName(nbt.getString("castername")) else setDead()
		if (caster == null) setDead()
	}
	
	public override fun writeEntityToNBT(nbt: NBTTagCompound) {
		if (caster is EntityPlayer) nbt.setString("castername", caster!!.commandSenderName)
	}
}